### PR TITLE
Add ordered dithering with custom color palettes

### PR DESCRIPTION
**Pull request by adrhill**
Implements the"Pattern Dithering" algorithm by Thomas Knoll which is/was used in Photoshop. [The patent appears to have expired in 2019.](https://patents.google.com/patent/US6606166B1/en)
___
I implemented the algorithm according to [Joel Yliluoma's pseudocode](https://bisqwit.iki.fi/story/howto/dither/jy/), however I'm using XYZ color space to add and substract colors and Lab color space to compute color differences.

I'm using Joel Yliluoma's test image to evaluate this implementation:
* Original image
  scene
* DitherPunk.jl (`color_error_multiplier=0.5`)
  scene_bayer_0 5
* Photoshop CS4
  scene_photoshop

Zooming in on the images, I feel like our implementation is slightly darker but has more accurate colors than Photoshop CS4.

The strength of the effect can be tweaked through the OrderedDither kwarg `color_error_multiplier`, which also affects the brightness of the output:
* `color_error_multiplier=1.0`
  scene_bayer_1 0
* `color_error_multiplier=0.5`
  scene_bayer_0 5
* `color_error_multiplier=0.1`
  scene_bayer_0 1

`color_error_multiplier=0` would be identical to `ClosestColor`.

This works for all `OrderedDithering` methods and all color schemes:
```Julia
using DitherPunk
using Clustering
using TestImages

img = testimage("peppers")
dither(img, ClusteredDots(), 5)
```
ClusteredDots_peppers_5

### CODE REVIEW

File: src/ordered.jl
Comment (johnnychen94) [inline src/ordered.jl:86]: How large `nmax` could be here? If it is too large then we can maybe try `nsmallest` and/or `nextreme` from DataStructures https://juliacollections.github.io/DataStructures.jl/latest/heaps/#Functions-using-heaps-1

This is a known algorithm "select kth smallest element" and is expected to have `O(n+klogn)` complexity (I'm not an algorithm expert, I googled it.)

Comment (johnnychen94) [reply src/ordered.jl:86]: Ah okay, it seems that `nmax=36` is quite small so that there's no need to apply fancy algorithms. But in place partial sort could still be useful.

```julia
julia> @btime partialsort!($x, 20) setup=(x=rand(1:100, 36));
  27.591 ns (0 allocations: 0 bytes)

julia> @btime sort($x) setup=(x=rand(1:100, 36));
  113.579 ns (1 allocation: 368 bytes)

julia> @btime sort!($x) setup=(x=rand(1:100, 36));
  73.069 ns (0 allocations: 0 bytes)
```

**Review by johnnychen94 — COMMENTED** [review]
I don't know the details but it looks fantastic! One small suggestion on the `sort` function.

**Review by johnnychen94 — APPROVED** [review]

File: src/ordered.jl
Comment (adrhill) [reply src/ordered.jl:86]: A reasonable size to benchmark on is `nmax=64`. In some exceptional cases, `nmax` could be really high
```julia
julia> prod(size(Bayer(level=4).mat))
1024
```
but this algorithm just isn't really well suited to such large threshold matrices.

Comment (adrhill) [reply src/ordered.jl:86]: Another approach is to precompute sorted colorschemes:
```julia
cs_lab = sort(Lab.(cs), by=c->c.l)
cs_xyz = XYZ.(cs_lab)
# ...
index[I] = sort(candidates)[mat[rlookup[r], clookup[c]]]
```
However, these indices then need to be mapped back to the original indices of `cs`.

File: src/ordered.jl
Comment (adrhill) [inline src/ordered.jl:82]: Breaking as soon as we are in a repeating subsequence would bring a huge speed-up.
In many cases, `candidates` ends up being something like `[1, 2, 1, 2, ...]` of length `nmax`.

**Review by johnnychen94 — COMMENTED** [review]
The failure on 1.0 is due to the fact that we're using an old `ColorTypes` version for Julia 1.0; `zero(XYZ)` requires ColorTypes v0.11. From CI output in #54, it is blocked by Images.

There are two possible solution, depending on how you want to solve it:

- (easy path): add compat patch to `compat.jl` for `zero`, e.g., `_zero(::XYZ) = XYZ{Float32}(0.0f0,0.0f0,0.0f0)`
- (hard path) replace Images with smaller packages in test dependency so that we don't need to hot-fix the version resolver in CI setup.

**Comment by johnnychen94** [comment]
It's funny to play with my own avatar

out

But yeah it's quite time-consuming

```julia
julia> out = @time dither(img, ClusteredDots(), 64)
┌ Warning: The clustering cost increased at iteration #69
└ @ Clustering ~/.julia/packages/Clustering/tt9vc/src/kmeans.jl:188
 27.842261 seconds (2.31 M allocations: 970.926 MiB, 0.64% gc time)
```

Before we consider adopting multi-threads, I feel we need to revisit the algorithm and find out an implementation with low complexity. Nevertheless, it's working!

File: src/ordered.jl
Comment (adrhill) [reply src/ordered.jl:82]: > It depends on how likely a repeat pattern occurs.

A repeat pattern is almost guaranteed! For `j=1`, `err = zero(XYZ)`. After that, for every iteration `j` from 2 to `nmax`
```julia
err = px - cs_xyz[idx]
```
so the number of possible values `err` takes on is _at most_ `1+length(cs)`. At `j=1+length(cs)`, we are therefore  guaranteed to be in a repeating subsequence. 

In practice, `idx = _closest_color_idx(...)` will only take on even fewer different numbers, and therefore also `err`. I'd guess we are very likely to be in a repeating subsequences around `j=4`.

**Comment by johnnychen94** [comment]
@adrhill 1.0 failures are fixed now. I'll leave all the rest to you; whether you want to eagerly merge this or merge after improving the performance.

**Comment by adrhill** [comment]
d8452d8 should bring a noticeable speedup. Benchmarking with
```julia
using ImageCore, DitherPunk, TestImages

img = testimage("fabio_color_256")
alg = Bayer()
cs = distinguishable_colors(20)

@benchmark dither($img, $alg, $cs)
```
I previously got 
```julia-repl
BenchmarkTools.Trial:
  memory estimate:  1.49 GiB
  allocs estimate:  89061301
  --------------
  minimum time:     4.176 s (1.72% GC)
  median time:      4.186 s (1.68% GC)
  mean time:        4.186 s (1.68% GC)
  maximum time:     4.195 s (1.65% GC)
  --------------
  samples:          2
  evals/sample:     1
```
and after the change:
```julia-repl
julia> @benchmark dither($img, $alg, $cs)
BenchmarkTools.Trial:
  memory estimate:  537.14 MiB
  allocs estimate:  29942237
  --------------
  minimum time:     1.406 s (1.50% GC)
  median time:      1.427 s (1.52% GC)
  mean time:        1.429 s (1.51% GC)
  maximum time:     1.457 s (1.63% GC)
  --------------
  samples:          4
  evals/sample:     1
```
The difference should be even more noticeable when using large palettes.

File: src/ordered.jl
Comment (adrhill) [inline src/ordered.jl:86]: Is there a more efficient way to fill the rest of the array with the loop `candidates[looprange]`?

Comment (adrhill) [reply src/ordered.jl:86]: When this is done, I will merge it so we can use the Benchmarking setup for further performance tweaks. 
I will leave our open conversations unresolved for easy referencing in the future as they contain some great ideas. 👍

Comment (johnnychen94) [reply src/ordered.jl:86]: One way is to use a lazy iterator, e.g., `Iterators.repeated`.

If you want, you can also bring a `repeat!` PR to julialang/julia repo; this should be low-hanging fruit and would be a great experience if you haven't done this before.

Comment (adrhill) [reply src/ordered.jl]: I went with `Iterators.take(Iterators.cycle(loop), n)`, however the performance seems unchanged.

**Pull request merged**
